eclair-core/src/main/scala/fr/acinq/eclair/wire/TlvTypes.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2019 ACINQ SAS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.wire
+
+import fr.acinq.eclair.UInt64
+import scodec.bits.ByteVector
+
+import scala.annotation.tailrec
+
+/**
+  * Created by t-bast on 20/06/2019.
+  */
+
+// @formatter:off
+trait Tlv {
+  val `type`: UInt64
+}
+sealed trait OnionTlv extends Tlv
+// @formatter:on
+
+/**
+  * Generic tlv type we fallback to if we don't understand the incoming type.
+  *
+  * @param `type` tlv type.
+  * @param value  tlv value (length is implicit, and encoded as a varint).
+  */
+case class GenericTlv(`type`: UInt64, value: ByteVector) extends Tlv
+
+/**
+  * A tlv stream is a collection of tlv records.
+  * A tlv stream is part of a given namespace that dictates how to parse the tlv records.
+  * That namespace is indicated by a trait extending the top-level tlv trait.
+  *
+  * @param records tlv records.
+  */
+case class TlvStream(records: Seq[Tlv]) {
+
+  // @formatter:off
+  sealed trait Error { val message: String }
+  case object RecordsNotOrdered extends Error { override val message = "tlv records must be ordered by monotonically-increasing types" }
+  case object DuplicateRecords extends Error { override val message = "tlv streams must not contain duplicate records" }
+  case object UnknownEvenTlv extends Error { override val message = "tlv streams must not contain unknown even tlv types" }
+  // @formatter:on
+
```

Comment:
These Error objects should be defined in TlvStream's companion object, not in the class itself

Reply:
Done.

Reply:
Why did you define your own `Error` class instead of reusing `scodec.Error`?

Reply:
How about:

```scala
case class TlvStream(records: Seq[Tlv]) {

  records.foldLeft(Option.empty[Tlv]) {
    case (None, record) => Some(record)
    case (Some(previousRecord), record) =>
      require(record.`type` == previousRecord.`type`, "tlv streams must not contain duplicate records")
      require(record.`type` > previousRecord.`type`, "tlv records must be ordered by monotonically-increasing types")
      require(record.`type`.toBigInt % 2 != 0, "tlv streams must not contain unknown even tlv types")
      Some(record)
  }
}
```

Codec would be:

```scala
def tlvStream(codec: Codec[Tlv]): Codec[TlvStream] = {
    Codec[TlvStream](
      (s: TlvStream) => {
        val recordTypes = s.records.map(_.`type`)
        if (recordTypes.length != recordTypes.distinct.length) {
          Attempt.Failure(scodec.Err("duplicate tlv records aren't allowed"))
        } else {
          Encoder.encodeSeq(codec)(s.records.sortBy(_.`type`).toList)
        }
      },
      (buf: BitVector) =>
        Decoder.decodeCollect[List, Tlv](codec, None)(buf).map(_.map(TlvStream(_)))
    )
  }
```